### PR TITLE
feat: object store to pvc migrations

### PR DESCRIPTION
**Pull request by DanStough**

### CODE REVIEW

File: src/markdown-pages/add-ons/registry.md
Comment (samantha-replicated) [inline src/markdown-pages/add-ons/registry.md:28]:
```suggestion
Migration can be performed on upgrade/re-install from an object store to a PV under the following circumstances:
```

File: src/markdown-pages/add-ons/kotsadm.md
Comment (samantha-replicated) [inline src/markdown-pages/add-ons/kotsadm.md:12]:
```suggestion
By default, this add-on requires an S3-Compatible object store be available in the cluster. 
```

File: src/markdown-pages/add-ons/kotsadm.md
Comment (samantha-replicated) [inline src/markdown-pages/add-ons/kotsadm.md:34]:
```suggestion
Object storage is used by the KOTS add-on as default behaviors in the following ways:
```

File: src/markdown-pages/add-ons/kotsadm.md
Comment (samantha-replicated) [inline src/markdown-pages/add-ons/kotsadm.md:36]:
```suggestion
1. Directly, when hostpath or NFS locations are set in the console. The console will dynamically create a Minio instance and attach it to the specified source.
```

File: src/markdown-pages/add-ons/kotsadm.md
Comment (samantha-replicated) [inline src/markdown-pages/add-ons/kotsadm.md:37]:
```suggestion
1. Indirectly, as part of the Registry add-on, to store container images in air gap deployments.
```

File: src/markdown-pages/add-ons/kotsadm.md
Comment (samantha-replicated) [inline src/markdown-pages/add-ons/kotsadm.md:38]:
```suggestion
1. Indirectly, as part of the Velero add-on, to store snapshots in the cluster using the `Internal` option.
```

File: src/markdown-pages/add-ons/kotsadm.md
Comment (samantha-replicated) [inline src/markdown-pages/add-ons/kotsadm.md:40]:
```suggestion
When the `disableS3` flag is set to `true`, the KOTS admin console will switch to use a PV for all of the direct operations. 
```

File: src/markdown-pages/add-ons/kotsadm.md
Comment (samantha-replicated) [inline src/markdown-pages/add-ons/kotsadm.md:41]:
```suggestion
It will also have the following side effects outside of the KOTS add-on:
```

File: src/markdown-pages/add-ons/kotsadm.md
Comment (samantha-replicated) [inline src/markdown-pages/add-ons/kotsadm.md:42]:
```suggestion
1. For new installs, the Registry add-on behavior will not change. It will use an object store if it is available, otherwise it will use a PV. For upgrades or re-installs, a migration will only be performed from an object store to a PV if the object store is removed from the installer spec.
```

File: src/markdown-pages/add-ons/kotsadm.md
Comment (samantha-replicated) [inline src/markdown-pages/add-ons/kotsadm.md:43]:
```suggestion
1. For new installs, the Velero add-on will use a PV for storing _Internal_ snapshots instead of using object storage. For upgrades or re-installs, _internal_ snapshots will be migrated from the object store to a PV.
```

File: src/markdown-pages/add-ons/kotsadm.md
Comment (samantha-replicated) [inline src/markdown-pages/add-ons/kotsadm.md:45]:
```suggestion
**Note:** The flag must be set to perform migrations.
```

File: src/markdown-pages/add-ons/registry.md
Comment (samantha-replicated) [inline src/markdown-pages/add-ons/registry.md:28]:
```suggestion
Migrations are performed on upgrade/re-install from an object store to a PV under the following circumstances.
```

File: src/markdown-pages/add-ons/velero.md
Comment (samantha-replicated) [inline src/markdown-pages/add-ons/velero.md:43]:
```suggestion
1. There is an object storage provider in the cluster. This could be satisfied by the Rook or Minio add-ons.
```

File: src/markdown-pages/add-ons/velero.md
Comment (samantha-replicated) [inline src/markdown-pages/add-ons/velero.md:45]:
```suggestion
You can also decide to move between these **Internal** storage locations. See [Removing Minio](/docs/install-with-kurl/removing-minio) for more information on this migration.
```

File: src/markdown-pages/add-ons/velero.md
Comment (samantha-replicated) [inline src/markdown-pages/add-ons/velero.md:47]:
```suggestion
**Warning:** None of these **Internal** locations are suitable for disaster recovery because loss of the cluster will mean loss of backups. 
```

File: src/markdown-pages/install-with-kurl/removing-minio.md
Comment (samantha-replicated) [inline src/markdown-pages/install-with-kurl/removing-minio.md:9]:
```suggestion
As of [v2021.12.20-0](https://kurl.sh/release-notes/v2021.12.20-0), kURL clusters can be installed without add-ons having a peer dependency on object storage. 
```

**Review by emosbaugh — COMMENTED** [review]
Do we prefer add-on or addon?

File: src/markdown-pages/install-with-kurl/removing-minio.md
Comment (samantha-replicated) [inline src/markdown-pages/install-with-kurl/removing-minio.md:10]:
```suggestion
There is also a migration path for clusters to remove object storage, similar to removing Minio.
```

File: src/markdown-pages/install-with-kurl/removing-minio.md
Comment (samantha-replicated) [inline src/markdown-pages/install-with-kurl/removing-minio.md:14]:
```suggestion
The following add-ons can be installed without object storage as noted:
```

File: src/markdown-pages/install-with-kurl/removing-minio.md
Comment (samantha-replicated) [inline src/markdown-pages/install-with-kurl/removing-minio.md:15]:
```suggestion
1. **Registry**: without an object store in the installer spec, a Persistent Volume (PV) will be used for storage. 

**Note:** The Registry will always use this as an object store if the Rook  add-on is installed.
```

File: src/markdown-pages/install-with-kurl/removing-minio.md
Comment (samantha-replicated) [inline src/markdown-pages/install-with-kurl/removing-minio.md:16]:
```suggestion
1. **Velero**: without object storage in the spec, no `default` BackupStorageLocation will be created. If the `disableS3` flag is set to `true` for the KOTS add-on, a PV-backed storage location will be created as the `default` location using the [Local-Volume-Provider](https://github.com/replicatedhq/local-volume-provider) plugin.
```

File: src/markdown-pages/install-with-kurl/removing-minio.md
Comment (samantha-replicated) [inline src/markdown-pages/install-with-kurl/removing-minio.md:49]:
```suggestion
Object storage can be removed from some of the add-ons in the cluster by simply setting the `kotsadm.disableS3` flag in the KOTS addon to `true`.
```

File: src/markdown-pages/install-with-kurl/removing-minio.md
Comment (samantha-replicated) [inline src/markdown-pages/install-with-kurl/removing-minio.md:50]:
```suggestion
See the documentation in the [KOTS add-on](/docs/add-ons/kotsadm) for more information on the `disableS3` flag.
```

File: src/markdown-pages/install-with-kurl/removing-minio.md
Comment (samantha-replicated) [inline src/markdown-pages/install-with-kurl/removing-minio.md:54]:
```suggestion
1. **Velero**: the default `BackupStorageLocation` will be updated to point at an attached Persistent Volume using the [Local-Volume-Provider](https://github.com/replicatedhq/local-volume-provider) plugin. A migration will be triggered to copy from the object store into the attached PV.
```

File: src/markdown-pages/install-with-kurl/removing-minio.md
Comment (samantha-replicated) [inline src/markdown-pages/install-with-kurl/removing-minio.md:60]:
```suggestion
In the case of Minio, it is a straightforward removal of the add-on.
```

File: src/markdown-pages/install-with-kurl/removing-minio.md
Comment (samantha-replicated) [inline src/markdown-pages/install-with-kurl/removing-minio.md:61]:
```suggestion
For clusters using the Rook addd-on, another CSI such as Longhorn or OpenEBS is required for storage. 
```

File: src/markdown-pages/install-with-kurl/removing-minio.md
Comment (samantha-replicated) [inline src/markdown-pages/install-with-kurl/removing-minio.md:63]:
```suggestion
Any object store data used by add-ons will also be migrated to PVs.
```

File: src/markdown-pages/install-with-kurl/removing-minio.md
Comment (samantha-replicated) [inline src/markdown-pages/install-with-kurl/removing-minio.md:65]:
```suggestion
See the documentation in the [KOTS add-on](/docs/add-ons/kotsadm) for more information on the `disableS3` flag.
```

**Review by samantha-replicated — CHANGES_REQUESTED** [review]
Hi Dan,

I did the first pass through of edits. I didn't commit my suggestions in case they conflicted with the technical accuracy of the doc. Please take a look and let me know if you have any questions. I wasn't sure if you were ready to merge this after review or if it is still a WIP. Keep me posted so I can update the review approvals so you can merge :)

**Comment by DanStough** [comment]
Thanks @samantha-replicated! Will review. Still WIP - also want Alex to review in case we need to tweak something in the language or in the actual product.

File: src/markdown-pages/add-ons/kotsadm.md
Comment (DanStough) [reply src/markdown-pages/add-ons/kotsadm.md]: resolved in comment by Samantha

Comment (DanStough) [reply src/markdown-pages/add-ons/kotsadm.md]: it's add-on

File: src/markdown-pages/install-with-kurl/removing-minio.md
Comment (DanStough) [reply src/markdown-pages/install-with-kurl/removing-minio.md:10]: Was going for something different here. I think I wanted to say removing Minio is a specific case of removing object storage. Suggestions on better wording appreciated.

File: src/markdown-pages/install-with-kurl/removing-minio.md
Comment (DanStough) [reply src/markdown-pages/install-with-kurl/removing-minio.md:15]: Can we keep this on the same line? I'm pretty sure the breaks will cause re-numbering of the list elements.

File: src/markdown-pages/install-with-kurl/removing-minio.md
Comment (DanStough) [reply src/markdown-pages/install-with-kurl/removing-minio.md:54]: `default` is the name name of the item and `BackupStorageLocation` is the type. I'll change them both to monospace.

File: src/markdown-pages/add-ons/kotsadm.md
Comment (ajp-io) [inline src/markdown-pages/add-ons/kotsadm.md:32]: I wonder if removing the word "Side" here would sound better. Side effects connote something unintentional, so I wonder if "Effects of the `disableS3` flag" would work better.

File: src/markdown-pages/add-ons/kotsadm.md
Comment (ajp-io) [inline src/markdown-pages/add-ons/kotsadm.md:41]: Same comment here, maybe "effects" instead of "side effects".

**Comment by ajp-io** [comment]
> Do we prefer add-on or addon?

I've been using it with a hyphen, which seems consistent for other things like Google Workspace add-ons and Firefox add-ons.

File: src/markdown-pages/add-ons/velero.md
Comment (ajp-io) [inline src/markdown-pages/add-ons/velero.md:39]: This isn't totally accurate anymore since we're supporting a PV now, right?

File: src/markdown-pages/install-with-kurl/removing-minio.md
Comment (samantha-replicated) [reply src/markdown-pages/install-with-kurl/removing-minio.md:15]:
```suggestion
1. **Registry**: without an object store in the installer spec, a Persistent Volume (PV) will be used for storage. **Note:** The Registry will always use this as an object store if the Rook add-on is installed.
```

Comment (samantha-replicated) [reply src/markdown-pages/install-with-kurl/removing-minio.md:15]: Good point! How about the above suggestion?

File: src/markdown-pages/add-ons/kotsadm.md
Comment (samantha-replicated) [reply src/markdown-pages/add-ons/kotsadm.md]: I like this suggestion!

File: src/markdown-pages/install-with-kurl/removing-minio.md
Comment (ajp-io) [reply src/markdown-pages/install-with-kurl/removing-minio.md:10]: What about this?
```suggestion
There is also a migration path for clusters to remove object storage, whether Rook or MinIO is used for object storage.
```

Comment (samantha-replicated) [reply src/markdown-pages/install-with-kurl/removing-minio.md:10]: I like this suggestion if that fits what you were trying to achieve Dan.

File: src/markdown-pages/install-with-kurl/removing-minio.md
Comment (ajp-io) [reply src/markdown-pages/install-with-kurl/removing-minio.md:15]: @DanStough Is this specific to Rook because people might not realize it comes with an object store?

Comment (DanStough) [reply src/markdown-pages/install-with-kurl/removing-minio.md:15]: I'm going to change `Rook` to object storage, since it covers both `Rook` and `Minio`, if that works for you?

File: src/markdown-pages/install-with-kurl/removing-minio.md
Comment (ajp-io) [inline src/markdown-pages/install-with-kurl/removing-minio.md:14]: Lots of add-ons can be installed without object storage. I like this wording since it shows that they can use object storage but now don't require it.
```suggestion
The following add-ons can now be installed with or without object storage:
```

File: src/markdown-pages/install-with-kurl/removing-minio.md
Comment (ajp-io) [inline src/markdown-pages/install-with-kurl/removing-minio.md:16]: Should default be in back-ticks or BackupStorageLocation? Is BackupStorageLocation a CR?

File: src/markdown-pages/install-with-kurl/removing-minio.md
Comment (ajp-io) [inline src/markdown-pages/install-with-kurl/removing-minio.md:43]: Could this confuse someone since "existing cluster" means something else in Replicated? Should we say pre-existing or something like that to disambiguate?

File: src/markdown-pages/install-with-kurl/removing-minio.md
Comment (ajp-io) [reply src/markdown-pages/install-with-kurl/removing-minio.md:15]: Yeah, that's fine.

File: src/markdown-pages/install-with-kurl/removing-minio.md
Comment (ajp-io) [inline src/markdown-pages/install-with-kurl/removing-minio.md:45]: Splitting this into two sections doesn't feel right to me. If you look at the three add-ons in each of the next two sections, we repeat a lot of info. For example, under removing the provider we talk about how KOTS will use a StatefulSet and migrate, but that isn't really because of the removal of the provider, that's because of the `disableS3` flag. I wonder if this would be clearer if we just had one section that talks about both actions (`disableS3: true` and removing the provider), and then for each action taken, we say which of those two is responsible for it. I can write it up, I just wanted to suggest it first to see if it makes more sense that way.

File: src/markdown-pages/install-with-kurl/removing-minio.md
Comment (DanStough) [reply src/markdown-pages/install-with-kurl/removing-minio.md:16]: `default` is the actual name of the object and BackupStorageLocation is the CR/type. @samantha-replicated what is the practice here?

File: src/markdown-pages/install-with-kurl/removing-minio.md
Comment (DanStough) [reply src/markdown-pages/install-with-kurl/removing-minio.md:43]: I like to defer to the simpler choice of `cluster migrations` if that works for you.

File: src/markdown-pages/install-with-kurl/removing-minio.md
Comment (DanStough) [reply src/markdown-pages/install-with-kurl/removing-minio.md:14]: Yeah, I definitely think I missed a word here. :thumbsup:

File: src/markdown-pages/install-with-kurl/removing-minio.md
Comment (ajp-io) [reply src/markdown-pages/install-with-kurl/removing-minio.md:43]: Works for me!

File: src/markdown-pages/install-with-kurl/removing-minio.md
Comment (DanStough) [reply src/markdown-pages/install-with-kurl/removing-minio.md:45]: I understand your point. I think my intent on making it into two sections was that there are two buckets of vendors in my mind:
1. Those that will enable this flag, but for some reason or another NOT remove Rook. Why? I don't know, but IICR some apps depend on the Rook RGW. Maybe this is also some migration path for them to mitigate a bunch of changes to their installer.
1. Those that will do everything in their power to remove Rook/Minio/Object Storage. 

If you would still like to change, I'm onboard.

File: src/markdown-pages/install-with-kurl/removing-minio.md
Comment (samantha-replicated) [reply src/markdown-pages/install-with-kurl/removing-minio.md:16]: General best practice is to back-tick objects and CR in-line.

File: src/markdown-pages/install-with-kurl/removing-minio.md
Comment (ajp-io) [inline src/markdown-pages/install-with-kurl/removing-minio.md:6]: @DanStough Is it okay that I changed these to the more generic "Removing Object Storage"? It felt like we were talking about removing object storage in general but then named this specifically after MinIO.

File: src/markdown-pages/install-with-kurl/removing-object-storage.md
Comment (ajp-io) [inline src/markdown-pages/install-with-kurl/removing-object-storage.md:45]: @DanStough I rewrote this so that each section only focused on what option only focuses on what that option does. Whereas before, the two sections repeated information because they mentioned stuff that the other migration path did. Does this read well to you?

File: src/markdown-pages/install-with-kurl/removing-minio.md
Comment (DanStough) [reply src/markdown-pages/install-with-kurl/removing-minio.md]: Yeah, I think that's fair. I picked MinIO because I thought that would be what folks would be immediately searching for, but this suggestion is a better general description.

File: src/markdown-pages/install-with-kurl/removing-object-storage.md
Comment (DanStough) [reply src/markdown-pages/install-with-kurl/removing-object-storage.md:45]: Nice! lgtm

**Comment by DanStough** [comment]
@replicatedhq/replicated-docs I think we are ready for a final review on this one.

File: src/markdown-pages/install-with-kurl/removing-object-storage.md
Comment (jonquil2002) [inline src/markdown-pages/install-with-kurl/removing-object-storage.md:52]: @DanStough What section is considered the "previous section for an example"? Are you referring to the YAML file under the New Installations section? It would be best to give a link to the specific topic/section you're referring to and some context. For example, you could say "When you re-install or upgrade using the updated installer spec (see the New Installations section for a YAML example), you should expect:..."

And add the link to the text "New Installations section"

File: src/markdown-pages/install-with-kurl/removing-object-storage.md
Comment (jonquil2002) [inline src/markdown-pages/install-with-kurl/removing-object-storage.md:64]: What section is considered the "previous section for an example"? Are you referring to the YAML file under the New Installations section? It would be best to give a link to the specific topic/section you're referring to and some context. For example, you could say "When you re-install or upgrade using the updated installer spec (see the New Installations section for a YAML example), you should expect:..."

And add the link to the text "New Installations section"

**Review by jonquil2002 — DISMISSED** [review]
LGTM but suggest making changes to the two comments that I left.

**Review by samantha-replicated — APPROVED** [review]
LGTM!

**Review by ajp-io — APPROVED** [review]

**Review by samantha-replicated — APPROVED** [review]

**Pull request merged**
